cppseminar/ApcScaler/ApcScaler/Workers.cs

```diff
@@ -0,0 +1,133 @@
+                throw new Exception("Missing configuration value of [AZURE_TENANT_ID]");
+            _ = !string.IsNullOrWhiteSpace(clientId) ? clientId :
+                throw new Exception("Missing configuration value of [AZURE_CLIENT_ID]");
+            _ = !string.IsNullOrWhiteSpace(clientSecret) ? clientSecret :
+                throw new Exception("Missing configuration value of [AZURE_CLIENT_SECRET]");
+
+            try
+            {
+                var creds = new AzureCredentialsFactory().FromServicePrincipal(clientId,
+                    clientSecret, tenantId, AzureEnvironment.AzureGlobalCloud);
+
+                _azureClient = Azure.Authenticate(creds).WithSubscription(subscriptionId);
+
+                _logger.LogDebug("Azure client successfully created");
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(Helpers.ExtractErrorMessage(e));
+            }
+        }
+        public int CountVm()
+        {
+            int allVmCounter = 0;
+            int runningVmCounter = 0;
+
+            _ = _azureClient ?? throw new Exception("_azureClient is null");
+            var vmInRg = _azureClient.VirtualMachines.ListByResourceGroup(_apcWorkerResourceGroup);
+            _logger.LogDebug("List of worker VMs successfully retrieved");
+
+            foreach (var virtualMachine in vmInRg)
+            {
+                if (virtualMachine.ComputerName.StartsWith(_apcWorkerVmNamePrefix))
+                {
+                    if (virtualMachine.PowerState == PowerState.Running)
+                    {
+                        runningVmCounter++;
+                    }
+
+                    if (virtualMachine.PowerState == PowerState.Stopped)
+                    {
+                        virtualMachine.Deallocate();
+                        _logger.LogInformation("Stopped worker VM (Name: {vmName}, Id:{vmId} was successfully deallocated",
+                            virtualMachine.ComputerName, virtualMachine.Id);
+                    }
+                }
+
+                allVmCounter++;
+            }
+
+            _logger.LogDebug("{allVmCounter} worker VM found ({runningVmCounter} running)", allVmCounter, runningVmCounter);
+
+            return runningVmCounter;
+        }
+        public void ScaleOutVm(int numberOfNewVm)
+        {
+            _logger.LogInformation("ScaleOut action started. We plan to start {numberOfNewVm} worker VMs.",
+                numberOfNewVm);
+
+            _ = _azureClient ?? throw new Exception("_azureClient is null");
+            var vmInRg = _azureClient.VirtualMachines.ListByResourceGroup(_apcWorkerResourceGroup);
+            _logger.LogDebug("List of worker VMs successfully retrieved");
+
+            foreach (var virtualMachine in vmInRg)
+            {
+                if (virtualMachine.ComputerName.StartsWith(_apcWorkerVmNamePrefix))
+                {
+                    if (virtualMachine.PowerState == PowerState.Deallocated)
```

Comment:
This is interesting, so we do not remove the VMs? If they are deallocated, they will be there in portal? Like not the worst thing in the world, but sill I do not like it very much. And what if we do not have VMs?

Reply:
Yes, idea is to (before app starts working) have deployed (to the given resource group) given max number (see MAX_NUMBER_OF_WORKERS setting in apcscaler app) of ready to work VMs in deallocated power state, and apcscaler app just starts and deallocate these VMs. VM is shutdown (go to power state stop) itself after some time of inactivity when they do not receive TestRun case. When VM is deallocated then we dot pay for it, and also starting VM is much more quicker than creating it (either from image or from scratch). We dot pay anything for deallocated VM so it seems to me that it is better to do it that way, than creating/deleting VMs all the way. Also while VMs are part of deployment and not created by app give us freedom to replace them if needed without disrupting/stopping app.